### PR TITLE
Optimize ImmutableQueue.CreateRange

### PR DESCRIPTION
**Pull request by jamesqo**
If the enumerable supports random access then we can push the items on to the forwards stack in reverse order. This avoids having to make an expensive allocation where (previously) when two items were dequeued, we would have to reverse the entire backwards stack.

If the enumerable does not support random access, we can still do better than the current implementation because we can deal with ImmutableStacks directly and not allocate an ImmutableQueue wrapper over them for every item.

/cc @ianhays, @AArnott, @safern

### CODE REVIEW

**Comment by jamesqo** [comment]
@dotnet-bot test code coverage please

**Comment by jamesqo** [comment]
@dotnet-bot test code coverage please

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue.cs [hunk 1]
```diff
@@ -34,27 +35,76 @@ public static ImmutableQueue<T> Create<T>()
...
 
-            return queue;
+            using (IEnumerator<T> e = items.GetEnumerator())
+            {
+                if (!e.MoveNext())
+                {
+                    return ImmutableQueue<T>.Empty;
+                }
+
+                var forwards = ImmutableStack.Create(e.Current);
```
Comment (ianhays) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue.cs]: Why not have `forwards` be empty and push all items on to `backwards`?

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue_1.cs [hunk 2]
```diff
@@ -151,7 +154,7 @@ public ImmutableQueue<T> Enqueue(T value)
 
             if (this.IsEmpty)
             {
-                return new ImmutableQueue<T>(ImmutableStack<T>.Empty.Push(value), ImmutableStack<T>.Empty);
+                return new ImmutableQueue<T>(ImmutableStack.Create(value), ImmutableStack<T>.Empty);
```
Comment (ianhays) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue_1.cs]: How is `ImmutableStack.Create(value)` better than `ImmutableStack<T>.Empty.Push(value)`?

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue_1.cs [hunk 3]
```diff
@@ -76,7 +75,11 @@ public ImmutableQueue<T> Clear()
         /// </value>
         public bool IsEmpty
         {
-            get { return _forwards.IsEmpty && _backwards.IsEmpty; }
+            get
+            {
+                Debug.Assert(!_forwards.IsEmpty || _backwards.IsEmpty);
```
Comment (ianhays) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue_1.cs]: Is this Debug assertion related to your other changes?

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue.cs (hunk 1 above)
Comment (jamesqo) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue.cs]: @ianhays We try to defer the allocation where we [reverse the backwards stack](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue_1.cs#L199) as much as possible. Pushing 1 item onto the forwards stack makes that take 2 Dequeues to happen instead of 1. (This is basically the existing behavior because [we push once](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue_1.cs#L154) onto the forwards stack if we're empty.)

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue_1.cs (hunk 3 above)
Comment (jamesqo) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue_1.cs]: I removed the `&& _backwards.IsEmpty` because AFAICS if the forwards stack is empty then the backwards is, too. I added this assert to ensure that.

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue_1.cs (hunk 2 above)
Comment (jamesqo) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue_1.cs]: The former is more concise and calls `ImmutableStack<T>.Empty.Push(value)`.

**Comment by jamesqo** [comment]
@dotnet-bot test code coverage please

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue.cs (hunk 1 above)
Comment (jamesqo) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue.cs]: @safern, I commented above.

Comment (safern) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue.cs]: @jamesqo yes I was reviewing and the comment didn't show until I've commented! Thanks

**Comment by stephentoub** [comment]
I'm not convinced this is a good idea. ```ImmutableList<T>``` is an ```IList<T>```, and IIRC, its indexer is not O(1), either O(log n) or O(n), I don't remember.  Especially since ```ImmutableList<T>``` and ```ImmutableQueue<T>``` are in the same library, it is not at all far fetched to envision someone creating the latter from the former, in which case this is changing something O(n) to something O(n log n) or O(n^2).

**Comment by jamesqo** [comment]
@stephentoub

> Especially since `ImmutableList<T>` and `ImmutableQueue<T>` are in the same library, it is not at all far fetched to envision someone creating the latter from the former,

OK. I can check for `ImmutableList<T>` before employing the optimization, which shouldn't hit perf too much since it's a sealed typecast. This will still be problematic for other non-O(1) indexer ILists like `LinkedList<T>`, but those should be very rare [and we don't seem to take those into account in Linq anyway](https://github.com/dotnet/corefx/blob/master/src/System.Linq/src/System/Linq/Last.cs#L116).

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue.cs [hunk 4]
```diff
@@ -34,47 +35,92 @@ public static ImmutableQueue<T> Create<T>()
...
         public static ImmutableQueue<T> CreateRange<T>(IEnumerable<T> items)
         {
             Requires.NotNull(items, nameof(items));
 
-            var queue = ImmutableQueue<T>.Empty;
-            foreach (var item in items)
+            var ilist = items as IList<T>;
+            // Even though ImmutableList supports random access, its indexer has O(log N) time complexity.
+            // So don't employ any IList-based optimizations if `items` is an ImmutableList.
+            if (ilist != null && !(items is ImmutableList<T>))
```
Comment (stephentoub) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue.cs]: Do we know of cases where CreateRange is used in performance-sensitive code paths with arbitrary ILists?  I'm trying to understand if this PR came about just from looking for things to change or if it was actually driven by a real-world need.  My previous comments about ImmutableList weren't meant to be interpreted as "special-case that one type" but rather "let's reconsider whether this optimization as a whole is really necessary".

Comment (jamesqo) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue.cs]: > Do we know of cases where CreateRange is used in performance-sensitive code paths with arbitrary ILists?

I would say that most inputs here will probably be ILists, but maybe you're correct that this is more of a convenience method rather than a function someone would use in a perf-sensitive code path. I'll remove the special-casing for IList since the indexer is a black box, and only specialize for arrays (there is a `Create(T[])` method, so we get that almost for free).

Comment (jamesqo) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue.cs]: @stephentoub BTW, however (I don't know if you saw it in the description) I only specialized for ILists in this case because it would avoid a big allocation when Dequeue reversed the backwards stack. The purpose of typecasting was not to reduce virtual method calls.

**Comment by jamesqo** [comment]
Test Innerloop Windows_NT Release Build and Test

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue.cs [hunk 5]
```diff
@@ -34,47 +35,69 @@ public static ImmutableQueue<T> Create<T>()
         [Pure]
         public static ImmutableQueue<T> Create<T>(T item)
         {
-            return ImmutableQueue<T>.Empty.Enqueue(item);
+            return new ImmutableQueue<T>(ImmutableStack.Create(item), ImmutableStack<T>.Empty);
```
Comment (stephentoub) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue.cs]: Why is this new version better?

**Comment by stephentoub** [comment]
Can you share the results of your perf tests?

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue.cs (hunk 5 above)
Comment (jamesqo) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue.cs]: @stephentoub This will be inlined while `ImmutableQueue.Enqueue` will not be because it has to handle the case where `this` is not empty. I'll post this in the benchmarks.

**Comment by jamesqo** [comment]
@stephentoub  https://gist.github.com/jamesqo/7452dfd29fe570b614a15c6785050d68

File: src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue.cs (hunk 5 above)
Comment (jamesqo) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableQueue.cs]: Hm, it looks like this is causing `Create` to not be inlined causing the perf difference to be nil.

**Comment by jamesqo** [comment]
@dotnet-bot test code coverage please

**Comment by jamesqo** [comment]
It seems like `System.Transactions.Tests` was hanging during a code coverage build: https://ci.dot.net/job/dotnet_corefx/job/master/job/code_coverage_windows_prtest/24/consoleFull#10606384068101fac1-935e-4556-95f9-7a0e25ca0b52 Going to open an issue.

**Comment by jamesqo** [comment]
@dotnet-bot test code coverage please

**Comment by danmosemsft** [comment]
@dotnet-bot test code coverage windows debug (https://github.com/dotnet/corefx/issues/16156)

**Comment by danmosemsft** [comment]
@dotnet-bot test code coverage windows debug (https://github.com/dotnet/corefx/issues/16156)

**Comment by jamesqo** [comment]
@dotnet-bot test code coverage please

**Comment by jamesqo** [comment]
Code coverage build is unblocked bc of https://github.com/dotnet/corefx/pull/16324, so trying again.

@dotnet-bot test code coverage please

**Comment by jamesqo** [comment]
Wow, got an AV this time (https://github.com/dotnet/corefx/issues/16371).

@dotnet-bot test code coverage please

**Comment by safern** [comment]
@stephentoub is this ready to go, or you are not convinced about this yet?

**Comment by jamesqo** [comment]
@safern It's me who's been slow on making progress here, actually. You can see from my comments above that I was trying to test code coverage here for 4 days and CI kept failing for different reasons. Now that the coverage report has finally succeeded, I'm seeing in https://ci.dot.net/job/dotnet_corefx/job/master/job/code_coverage_windows_prtest/32/Code_Coverage_Report/ that the case where a non-array is passed to `CreateRange` is never covered. I'll have to add a test for that before this PR can get merged.

**Comment by safern** [comment]
@jamesqo sounds good! Thanks for the update!

**Comment by jamesqo** [comment]
K, I added an extra test in the latest commit and code coverage is now at 100% for the `ImmutableQueue` factory class.

**Comment by jamesqo** [comment]
Test Innerloop Windows_NT Debug Build and Test (https://github.com/dotnet/corefx/issues/16651)

**Comment by jamesqo** [comment]
@stephentoub Does this PR seem fine to merge? I've addressed your concerns above, shared the perf tests, and added an extra test to make sure all branches are being covered.
